webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/services/impl/GitTechniqueReader.scala

```diff
@@ -280,21 +276,49 @@ class GitTechniqueReader(
         //from other case (just simple updates)
 
         if(versionsMods.values.forall( _ == VersionDeleted)
-           && currentTechniquesInfoCache.techniques.get(name).map( _.size) == Some(versionsMods.size)
+           && currentTechniquesInfo.techniques.get(name).map( _.size) == Some(versionsMods.size)
         ) {
           (name, TechniqueDeleted(name, versionsMods.keySet))
         } else {
           (name, TechniqueUpdated(name, versionsMods))
         }
       }.toMap
-
-      //Ok, now rebuild Technique !
-      modifiedTechniquesCache = techniqueMods
-      nextTechniquesInfoCache = (nextId, nextTechniquesInfo)
-      modifiedTechniquesCache
     }
+
+    (for {
+      nextId   <- revisionProvider.getAvailableRevTreeId
+      cached   <- nextTechniquesInfoCache.get
+      db       <- repo.db
+      mods     <- if(nextId == cached._1) modifiedTechniquesCache.get
+                  else for {
+                    nextTechniquesInfo <- processRevTreeId(db, nextId)
+                    managedDiffFmt     =  ZManaged.make(IOResult.effect{
```

Comment:
`ZManaged.make(open resources)(close resources)` allows to create a safe manager for a resources. You give it an `open` and `close` logic, and then you safely use the resource with `managedThing.use(resources => ...)`. It's typically the good option from `io streams`, and git related things like `DiffFormatter` and `TreeWalk` (below).